storage/local/doubledelta.go

```diff
@@ -0,0 +1,513 @@
+	default:
+		panic("Invalid number of bytes for time delta")
+	}
+
+	offset += int(c.timeBytes())
+
+	var v clientmodel.SampleValue
+	if c.isInt() {
+		switch c.valueBytes() {
+		case d0:
+			v = c.baseValue() +
+				clientmodel.SampleValue(idx)*c.baseValueDelta()
+		case d1:
+			v = c.baseValue() +
+				clientmodel.SampleValue(idx)*c.baseValueDelta() +
+				clientmodel.SampleValue(int8(c[offset]))
+		case d2:
+			v = c.baseValue() +
+				clientmodel.SampleValue(idx)*c.baseValueDelta() +
+				clientmodel.SampleValue(int16(binary.LittleEndian.Uint16(c[offset:])))
+		case d4:
+			v = c.baseValue() +
+				clientmodel.SampleValue(idx)*c.baseValueDelta() +
+				clientmodel.SampleValue(int32(binary.LittleEndian.Uint32(c[offset:])))
+		// No d8 for ints.
+		default:
+			panic("Invalid number of bytes for integer delta")
+		}
+	} else {
+		switch c.valueBytes() {
+		case d4:
+			v = c.baseValue() +
+				clientmodel.SampleValue(idx)*c.baseValueDelta() +
+				clientmodel.SampleValue(math.Float32frombits(binary.LittleEndian.Uint32(c[offset:])))
+		case d8:
+			// Take absolute value for d8.
+			v = clientmodel.SampleValue(math.Float64frombits(binary.LittleEndian.Uint64(c[offset:])))
+		default:
+			panic("Invalid number of bytes for floating point delta")
+		}
+	}
+	return &metric.SamplePair{
+		Timestamp: ts,
+		Value:     v,
+	}
+}
+
+// firstTime implements chunk.
+func (c doubleDeltaEncodedChunk) firstTime() clientmodel.Timestamp {
+	return c.baseTime()
+}
+
+// lastTime implements chunk.
+func (c doubleDeltaEncodedChunk) lastTime() clientmodel.Timestamp {
+	return c.valueAtIndex(c.len() - 1).Timestamp
+}
+
+// marshal implements chunk.
+func (c doubleDeltaEncodedChunk) marshal(w io.Writer) error {
+	if len(c) > math.MaxUint16 {
+		panic("chunk buffer length would overflow a 16 bit uint.")
+	}
+	binary.LittleEndian.PutUint16(c[doubleDeltaHeaderBufLenOffset:], uint16(len(c)))
+
+	n, err := w.Write(c[:cap(c)])
+	if err != nil {
+		return err
+	}
+	if n != cap(c) {
+		return fmt.Errorf("wanted to write %d bytes, wrote %d", len(c), n)
+	}
+	return nil
+}
+
+// unmarshal implements chunk.
+func (c *doubleDeltaEncodedChunk) unmarshal(r io.Reader) error {
+	*c = (*c)[:cap(*c)]
+	readBytes := 0
+	for readBytes < len(*c) {
+		n, err := r.Read((*c)[readBytes:])
+		if err != nil {
+			return err
+		}
+		readBytes += n
+	}
+	*c = (*c)[:binary.LittleEndian.Uint16((*c)[doubleDeltaHeaderBufLenOffset:])]
+	return nil
+}
+
+// doubleDeltaEncodedChunkIterator implements chunkIterator.
+type doubleDeltaEncodedChunkIterator struct {
+	chunk *doubleDeltaEncodedChunk
+	// TODO(beorn7): add more fields here to keep track of last position.
+}
+
+// newIterator implements chunk.
+func (c *doubleDeltaEncodedChunk) newIterator() chunkIterator {
+	return &doubleDeltaEncodedChunkIterator{
+		chunk: c,
+	}
+}
+
+// getValueAtTime implements chunkIterator.
+func (it *doubleDeltaEncodedChunkIterator) getValueAtTime(t clientmodel.Timestamp) metric.Values {
```

Comment:
(that would also enable us to make `chunk.values()` a function independent of the chunk type)